### PR TITLE
[frameit] Support multiple cascading filters

### PR DESCRIPTION
**Pull request by bvirlet**
🔑 

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

My existing #11014 PR didn't get merged – admittedly, it wasn't easy to review – and got stale. However, it mixed different changes and some of them have been already addressed now (I believe) by #11585.

This part hasn't been addressed though, so I would like to resubmit if for review:

#### Add cascading filters

You may want to adjust the layout not only based on the screenshot number/name, but also on the device. Currently, only the first filter matching the screenshot path is applied. With this change, the filters are applied in order.

This lets modify the iPad screenshot without modifying the iPhone's:

```
{
  "device_frame_version": "latest",
  "default": {
    "title": {
      "color": "#ffffff"
    },
    "background": "./Frameit-background.png",
    "padding": "3%",
    "show_complete_frame": false,
    "stack_title" : true,
    "title_below_image": false,
    "font_scale_factor": 0.05
  },
   "data": [
    {
      "filter": "scan"
    },
    {
      "filter": "edit"
    },
    {
      "filter": "multipage_pdf"
    },
    {
      "filter": "organize"
    },
    {
      "filter": "export"
    },
    {
      "filter" : "iPad",
      "show_complete_frame" : true,
      "font_scale_factor": 0.035
    }
  ]
}
```

### Documentation

Regarding the documentation update, here is the proposed change:

Before:
```
This is mandatory to link the individual configuration to the screenshot, based on part of the file name.
Example:
If a screenshot is named iPhone 8-Brainstorming.png you can use value Brainstorming for filter. All other keys from that array element will only be applied on this specific screenshot.
```
After:
```
This is mandatory to link the individual configuration to the screenshot, based on part of the file name.
Example:
If a screenshot is named `iPhone5_Brainstorming.png` the first entry in the `data` array will be used. If there are more than one `filter` matching an entry, they will all be applied in order (which means that the last one has the highest precedence).
```
If I understand correctly, I will need to make the corresponding PR to the fastlane/docs repo?

### CODE REVIEW

**Comment by joshdholtz** [comment]
@bvirlet Heyyy! TBH, I don't have frameit setup to test this PR but I will get something setup to test this 💪

**Comment by bvirlet** [comment]
Hi @joshdholtz ? Do you still you will find some time to review this? I think it can be really useful to others but I'm afraid this might go stale again 😄  Thanks for all what you are doing for fastlane in any case!

**Comment by janpio** [comment]
Hey @bvirlet, could you maybe explain the code change in plain English? Right now I don't really get it :/

> If I understand correctly, I will need to make the corresponding PR to the fastlane/docs repo?

Fortunately not, you can just edit this file in this repository: https://github.com/fastlane/fastlane/blob/master/fastlane/lib/fastlane/actions/docs/frame_screenshots.md

**Comment by bvirlet** [comment]
hi @janpio. Thanks for taking the time to check this.

Line 29: instead of just finding the first `data` object of the JSON matching the filter, we find _all_ of them.
Line 33-36: before, we were merging the selected `data` object of the JSON into the default data object. Now we merge all the selected ones. We have to clone the `default` object because otherwise we would modify the object we are iterating on (that wasn't a problem before with just a single object).
A good way to see what happens would be to just print the result before and after line 36.

**Comment by janpio** [comment]
Ah ok, I wasn't aware that it was just searching for any match for the `filter` string - I thought it had to be an exact match of the non-device part. This doesn't seem to be the case though.

---

BUT: It seems this PR is already merged via the rebased https://github.com/fastlane/fastlane/pull/12484! Correct @bvirlet? https://docs.fastlane.tools/actions/frameit/#specific-parameters also already has your wording from your original description here, correct?

That being said: We should really improve the documentation and make sure this is all written down somewhere...

**Review by janpio — CHANGES_REQUESTED** [review]
Don't merge: Already part of the code base via #12484

**Comment by bvirlet** [comment]
Ah yes, looks like @futuretap rebased it and got it merged in the meantime!

**Comment by janpio** [comment]
Feel free to create new PRs, next time I will make sure they get merged a "bit" faster 🎂

**Pull request closed without merging**
